homeassistant/components/media_player/heos.py

```diff
@@ -0,0 +1,217 @@
+"""Denon HEOS Media Player."""
+import asyncio
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.media_player import (PLATFORM_SCHEMA,
+                                                   MediaPlayerDevice)
+
+from homeassistant.components.media_player.const import (
+    MEDIA_TYPE_MUSIC, SUPPORT_NEXT_TRACK, SUPPORT_PAUSE, SUPPORT_PLAY,
+    SUPPORT_PLAY_MEDIA, SUPPORT_PREVIOUS_TRACK, SUPPORT_SEEK, SUPPORT_STOP,
+    SUPPORT_VOLUME_MUTE, SUPPORT_VOLUME_SET, SUPPORT_VOLUME_STEP)
+
+from homeassistant.const import (CONF_HOST, CONF_NAME, CONF_PASSWORD,
+                                 CONF_USERNAME, STATE_OFF, STATE_PAUSED,
+                                 STATE_PLAYING, STATE_UNKNOWN)
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['aioheos==0.1.5']
+
+DEFAULT_NAME = 'HEOS Player'
+
+SUPPORT_HEOS = SUPPORT_PLAY | SUPPORT_STOP | SUPPORT_PAUSE | \
+        SUPPORT_PLAY_MEDIA | SUPPORT_PREVIOUS_TRACK | SUPPORT_NEXT_TRACK | \
+        SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_SET | SUPPORT_VOLUME_STEP | \
+        SUPPORT_SEEK
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_HOST):
+    cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME):
+    cv.string,
+    vol.Optional(CONF_USERNAME):
+    cv.string,
+    vol.Optional(CONF_PASSWORD):
+    cv.string,
+})
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discover_info=None):
+    """Set up the HEOS platform."""
+    host = config.get(CONF_HOST)
+    name = config.get(CONF_NAME)
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+
+    heos = HeosMediaPlayer(hass, host, name, username, password)
+
+    yield from heos.heos.connect(
+        host=host, callback=heos.async_update_ha_state)
```

Comment:
We shouldn't directly tie the callback to `async_update_ha_state` as it needs to consider if the platform has been added to HASS.  Use the dispatcher (`hass.helpers.dispatcher`) instead:
- Create a callback that when invoked calls `hass.helpers.dispatcher.async_dispatcher_send(...)`
- In the platform, implement `async_added_to_hass` to connect a callback to the dispatcher using `hass.helpers.dispatcher.async_dispatcher_connect(...)` that calls `async_update_ha_state`. [See this example.](https://github.com/home-assistant/home-assistant/blob/master/homeassistant/components/smartthings/__init__.py#L332)
- Implment `async_will_remove_from_hass` to remove the connection to the dispatcher.
- [Dispatcher notes in the dev docs checklist](https://developers.home-assistant.io/docs/en/creating_component_code_review.html#3-component-platform-communication)

Reply:
Did some change here as well. I hope I understand it correctly.